### PR TITLE
custom filter classes

### PR DESCRIPTION
**Pull request by Mahlet2123**
More organized and maintainable code
​
## Description

1. Creating custom filter classes can make the code more organized and maintainable.
2. Defined custom exception classes for different error scenarios for the 'products-filter/' endpoint. This makes error handling more organized and allows  to provide clear and consistent error messages.
3. Documentation: providing clear information about the available filters, query parameters, and expected responses.

## How Has This Been Tested?

Tested Using Postman
​
## Screenshots (if appropriate - Postman, etc):
​image

Custom exception classes: 
eg -> InvalidFilterParams

In this image I intered an invalid price parameter and it is handled by the exception handler as shown
image
but here with the appropriate parameters
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
​
## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: product_filter/views.py [hunk 1]
```diff
@@ -7,18 +7,95 @@
 from django.http import Http404
 from rest_framework.pagination import PageNumberPagination
 
+# Custom Filters: Custom filtering classes.
+class KeywordFilter:
+    def filter(self, queryset, keywords):
```
Comment (zxenonx) [inline product_filter/views.py]: Kindly add type hint for keywords

File: product_filter/views.py [hunk 2]
```diff
@@ -7,18 +7,95 @@
...
+        if keywords:
+            return queryset.filter(Q(name__icontains=keywords) | Q(description__icontains=keywords) | Q(category__name__icontains=keywords))
+        return queryset
+
+class CategoryFilter:
+    def filter(self, queryset, category):
```
Comment (zxenonx) [inline product_filter/views.py]: type hint for category

File: product_filter/views.py [hunk 3]
```diff
@@ -7,18 +7,95 @@
...
+        return queryset.filter(category__name=category) if category else queryset
+
+class SubCategoryFilter:
+    def filter(self, queryset, sub_category):
```
Comment (zxenonx) [inline product_filter/views.py:22]: type hint for sub_category

File: product_filter/views.py [hunk 4]
```diff
@@ -7,18 +7,95 @@
...
+        return queryset.filter(category__productsubcategory__name=sub_category) if sub_category else queryset
+    
+class DiscountFilter:
+    def filter(self, queryset, discount):
```
Comment (zxenonx) [inline product_filter/views.py]: type hint for discount

File: product_filter/views.py [hunk 5]
```diff
@@ -7,18 +7,95 @@
...
+        return queryset.filter(discount_price=discount) if discount else queryset
+    
+class PriceFilter:
+    def filter(self, queryset, min_price, max_price):
```
Comment (zxenonx) [inline product_filter/views.py]: type hints for min_price, max_price

File: product_filter/views.py [hunk 6]
```diff
@@ -7,18 +7,95 @@
...
+        return queryset.filter(Q(price__gte=min_price) & Q(price__lte=max_price)) if min_price and max_price else queryset
 
-# Create your views here.
-class FilterProductView(APIView):
-
-    # pagination_class = PageNumberPagination
-    # page_size = 10
+class RatingFilter:
+    def filter(self, queryset, rating):
```
Comment (zxenonx) [inline product_filter/views.py]: type hint for rating

File: product_filter/views.py [hunk 7]
```diff
@@ -7,18 +7,95 @@
...
+class HighestPriceFilter:
+    def filter(self, queryset, highest_price):
+        if highest_price:
+            if highest_price == 'true':
+                # Implement logic to order by highest price
+                return queryset.order_by('-price')
+            else:
+                # Implement logic to order by lowest price
+                return queryset.order_by('price')
+        return queryset
```
Comment (zxenonx) [inline product_filter/views.py]: We have to refactor this function.
What's the type of highest_price?

File: product_filter/views.py [hunk 8]
```diff
@@ -7,18 +7,95 @@
...
+        - 400 Bad Request: If the request contains invalid filter parameters.
+        - 404 Not Found: If there are no matching products.
+        - 500 Internal Server Error: For any other unexpected errors.
+
+        Example Request:
+        GET /api/marketplace/products-filter/?category=Electronics&discount=1000.00&keywords=laptop
+
+        Example Response:
+        {
+            "success": true,
```
Comment (zxenonx) [inline product_filter/views.py:71]: Good.

**Review by zxenonx — CHANGES_REQUESTED** [review]
Good job @Mahlet2123 
I left some comments

File: product_filter/views.py (hunk 7 above)
Comment (Mahlet2123) [reply product_filter/views.py]: yes, I can make the logic more concise

If the "highest_price" parameter is set to "true," it sorts the products from the highest price to the lowest price. 
If set to "false", it sorts them from the lowest price to the highest price. 

If "highest_price" is not provided the query set remains unsorted.

**Review by zxenonx — COMMENTED** [review]
Thanks @Mahlet2123

**Comment by zxenonx** [comment]
```

  def filter(self, queryset, highest_price: str):
        if not highest_price:
            return queryset
        if highest_price != 'true':
            return queryset.order_by('price')
        return queryset.order_by('-price')
                
       @Mahlet2123 
```

**Comment by Mahlet2123** [comment]
> ```
> 
>   def filter(self, queryset, highest_price: str):
>         if not highest_price:
>             return queryset
>         if highest_price != 'true':
>             return queryset.order_by('price')
>         return queryset.order_by('-price')
>                 
>        @Mahlet2123 
> ```

yeah, this one is a more conscise logic

**Comment by zxenonx** [comment]
> > ```
> > 
> >   def filter(self, queryset, highest_price: str):
> >         if not highest_price:
> >             return queryset
> >         if highest_price != 'true':
> >             return queryset.order_by('price')
> >         return queryset.order_by('-price')
> >                 
> >        @Mahlet2123 
> > ```
> 
> yeah, this one is a more conscise logic

Alright

**Comment by Mahlet2123** [comment]
@zxenonx, Commited all the necessary changes requested.

**Review by zxenonx — APPROVED** [review]
Thanks  @Mahlet2123

**Pull request merged**